File: onboard/DroneBrain/app/src/main/java/edu/cmu/cs/dronebrain/MainActivity.java
package edu.cmu.cs.dronebrain;

import android.app.Activity;
import android.content.Context;
import android.location.Location;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.StrictMode;
import android.util.Log;
import android.view.WindowManager;
import android.webkit.URLUtil;

import com.google.protobuf.Any;
import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.parrot.drone.groundsdk.ManagedGroundSdk;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.DatagramSocket;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.SocketException;
import java.net.URL;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.function.Consumer;

import javax.net.SocketFactory;

import dalvik.system.DexClassLoader;
import edu.cmu.cs.dronebrain.impl.DebugCloudlet;
import edu.cmu.cs.dronebrain.impl.parrotanafi.ParrotAnafi;
import edu.cmu.cs.dronebrain.interfaces.CloudletItf;
import edu.cmu.cs.dronebrain.interfaces.DroneItf;
import edu.cmu.cs.dronebrain.interfaces.FlightScript;
import edu.cmu.cs.dronebrain.interfaces.Platform;
import edu.cmu.cs.gabriel.client.comm.ServerComm;
import edu.cmu.cs.gabriel.client.results.ErrorType;
import edu.cmu.cs.gabriel.protocol.Protos.InputFrame;
import edu.cmu.cs.gabriel.protocol.Protos.ResultWrapper;
import edu.cmu.cs.gabriel.protocol.Protos.PayloadType;
import edu.cmu.cs.steeleagle.Protos;
import edu.cmu.cs.steeleagle.Protos.Extras;


public class MainActivity extends Activity implements Consumer<ResultWrapper> {
    private ManagedGroundSdk sdk = null;
    /** Connectivity Manager to retrieve network infos **/
    private ConnectivityManager cm = null;
    /** LTE network object **/
    private Network LTEnetwork = null;
    /** Currently executing FlightScript **/
    private FlightScript MS = null;
    /** Reference to the thread executing the FlightScript **/
    private Thread scriptRunnable = null;
    /** Current drone object **/
    private DroneItf drone = null;
    /** Current cloudlet object **/
    private CloudletItf cloudlet = null;
    /** For Gabriel client connection **/
    private ServerComm serverComm = null;
    /** Script URL send from commander **/
    private String scriptUrl = null;
    private Handler loopHandler = null;
    private int heartbeatsSent = 0;
    /** Log tag **/
    String TAG = "DroneBrain";
    String SOURCE = "command";


    // Based on
    // https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/compiler/java/java_message.cc#L1387
    private static Any pack(Extras engineFields) {
        return Any.newBuilder()
                .setTypeUrl("type.googleapis.com/cnc.Extras")
                .setValue(engineFields.toByteString())
                .build();
    }

    public String getWifiSSID() {
        WifiManager wm = (WifiManager) this.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        if(wm != null) {
            return wm.getConnectionInfo().getSSID();
        } else {
            return "";
        }
    }

    @Override
    public void accept(ResultWrapper resultWrapper) {
        if (resultWrapper.getResultsCount() != 1) {
            Log.e(TAG, "Got " + resultWrapper.getResultsCount() + " results in output.");
            return;
        }

        ResultWrapper.Result result = resultWrapper.getResults(0);
        try {
            Extras extras = Extras.parseFrom(resultWrapper.getExtras().getValue());
            ByteString r = result.getPayload();
            if(extras.hasCmd()) {
                Protos.Command cmd = extras.getCmd();
                if(cmd.getHalt()) {
                    Log.i(TAG, "Killswitch signaled from commander.");
                    if (scriptRunnable != null) {
                        scriptRunnable.interrupt();
                        Log.i(TAG, "Interrupting script thread.");
                    }
                    if (drone != null) {
                        Log.i(TAG, "Killing drone.");
                        drone.kill();
                    }
                    finish();
                }

                if(URLUtil.isValidUrl(cmd.getScriptUrl())) {
                    Log.i(TAG, "Flight script sent by commander.");
                    scriptUrl = cmd.getScriptUrl();
                    Log.i(TAG, scriptUrl);
                    if(retrieveFlightScript())
                        executeFlightScript();
                }
            }
        } catch (InvalidProtocolBufferException e) {
            Log.e(TAG, "Protobuf Error", e);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (result.getPayloadType() != PayloadType.TEXT) {
            Log.e(TAG, "Got result of type " + result.getPayloadType().name());
            return;
        }

    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
        StrictMode.setThreadPolicy(policy);

        sdk = ManagedGroundSdk.obtainSession(this);

        CountDownLatch countDownLatch = new CountDownLatch(1);
        /** Create LTE network binding **/
        cm = (ConnectivityManager) getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkRequest.Builder req = new NetworkRequest.Builder();
        req.addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR);
        cm.requestNetwork(req.build(), new ConnectivityManager.NetworkCallback() {
                    @Override
                    public void onAvailable(Network network) {
                        /** Assign LTE network object **/
                        LTEnetwork = network;
                        countDownLatch.countDown();
                    }
                }
        );
        try {
            countDownLatch.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        Consumer<ErrorType> onDisconnect = errorType -> {
            Log.e(TAG, "Disconnect Error:" + errorType.name());
            finish();
        };

        serverComm = ServerComm.createServerComm(
                this, BuildConfig.GABRIEL_HOST, BuildConfig.PORT, getApplication(), onDisconnect);

        loopHandler = new Handler();
        loopHandler.postDelayed(gabrielLoop, 1000);

        /** Bind all future sockets to Wifi **/
        //bindProcessToWifi();
    }

    public Network getNetworkObjectForCurrentWifiConnection() {
        List<Network> networks = Arrays.asList(cm.getAllNetworks());
        for (Network network : networks) {
            NetworkCapabilities capabilities = cm.getNetworkCapabilities(network);
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                return network;
            }
        }
        return null;
    }

    private void bindProcessToWifi() {
        Network network = getNetworkObjectForCurrentWifiConnection();
        if (network != null) {
            Boolean success = cm.bindProcessToNetwork(network);
            Log.d("WifiConnect", success.toString());
        }
    }

    private Runnable gabrielLoop = new Runnable() {

        @Override
        public void run() {
            heartbeatsSent++;
            serverComm.sendSupplier(() -> {
                String p = "heartbeat";

                Extras extras;
                Extras.Builder extrasBuilder = Extras.newBuilder();
                Protos.Location.Builder lb = Protos.Location.newBuilder();
                Location l = null;
                try {
                    if(drone != null) {
                        lb.setLatitude(drone.getLat());
                        lb.setLongitude(drone.getLon());
                        extrasBuilder.setLocation(lb);
                    } else {
                        lb.setLatitude(0);
                        lb.setLongitude(0);
                        extrasBuilder.setLocation(lb);
                    }

                    extrasBuilder.setDroneId(Build.ID);

                    if(heartbeatsSent < 2) {
                        extrasBuilder.setRegistering(true);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                extras = extrasBuilder.build();

                return InputFrame.newBuilder()
                        .setPayloadType(PayloadType.TEXT)
                        .addPayloads(ByteString.copyFromUtf8(p))
                        .setExtras(pack(extras))
                        .build();
            }, SOURCE, false);
            loopHandler.postDelayed(this, 1000);
        }
    };

    protected boolean retrieveFlightScript() {
        boolean success = false;
        /** Download flight plan **/
        File f = new File(getDir("dex", MODE_PRIVATE), "classes.dex");
        try {
            Log.d(TAG, "Starting flight plan download");
            download(scriptUrl, f);
            /** Read data from flight plan **/
            DexClassLoader classLoader = new DexClassLoader(
                    f.getAbsolutePath(), getDir("outdex", MODE_PRIVATE).getAbsolutePath(),
                    null, getClassLoader());
            Class clazz = classLoader.loadClass("edu.cmu.cs.dronebrain.MS");
            Log.i(TAG, clazz.toString());
            MS = (FlightScript)clazz.newInstance();
            Log.d(TAG, "Getting drone and cloudlet...");
            drone = getDrone(MS.platform); // Get the corresponding drone
            cloudlet = getCloudlet(); // Get the corresponding cloudlet
            success = true;
        } catch (Exception e) {
            Log.e(TAG, "Download failed! " + e.toString());
        }
        return success;
    }

    protected void executeFlightScript() {
        try {
            /** Execute flight plan **/
            Log.d(TAG, "Executing flight plan!");
            MS.setDrone(drone);
            MS.setCloudlet(cloudlet);
            scriptRunnable = new Thread(new Runnable() {
                @Override
                public void run() {
                    MS.run();
                }
            });
            scriptRunnable.start();
        } catch (Exception e) {
            Log.e(TAG, e.toString());
        }
    }

    @Override
    protected void onStart(){
        super.onStart();


    }

    @Override
    protected void onResume() {
        super.onResume();
        heartbeatsSent = 0;
    }

    private DroneItf getDrone(Platform platform) throws Exception {
        if (platform == Platform.ANAFI) {
            return new ParrotAnafi(sdk);
        } else {
            throw new Exception("Unsupported platform: " + platform.name());
        }
    }

    private CloudletItf getCloudlet() {
        return new DebugCloudlet();
    }

    /** This needs to change in future to pull from CNC module **/
    private void download(String url, File dex) throws Exception {
        BufferedInputStream bis;
        BufferedOutputStream bos;
        try {
            URL urlObject = new URL(url);
            HttpURLConnection httpConn = (HttpURLConnection) LTEnetwork.openConnection(urlObject);
            int responseCode = httpConn.getResponseCode();
            if (responseCode != HttpURLConnection.HTTP_OK) {
                throw new RuntimeException("Download didn't work, http status: " + responseCode);
            }
            bis = new BufferedInputStream(httpConn.getInputStream());
            bos = new BufferedOutputStream(new FileOutputStream(dex));
            copy(bis, bos);
            bis.close();
            bos.close();
        } catch (IOException e) {
            throw new IOException(e);
        }
    }

    /** Helper function used by download **/
    private void copy(BufferedInputStream src, OutputStream dst) throws IOException {
        int len;
        byte[] buf = new byte[8 * 1024];
        while ((len = src.read(buf, 0, 8 * 1024)) > 0) {
            dst.write(buf, 0, len);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        sdk.close();
    }
}

File: onboard/DroneBrain/app/src/main/java/edu/cmu/cs/dronebrain/MainActivity.kt
package edu.cmu.cs.dronebrain

import android.app.Activity
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Bundle
import android.os.StrictMode
import android.os.StrictMode.ThreadPolicy
import android.util.Log
import android.widget.Button
import android.widget.TextView
import com.parrot.drone.groundsdk.GroundSdk
import com.parrot.drone.groundsdk.ManagedGroundSdk
import com.parrot.drone.groundsdk.Ref
import com.parrot.drone.groundsdk.device.DeviceState
import com.parrot.drone.groundsdk.device.Drone
import com.parrot.drone.groundsdk.device.RemoteControl
import com.parrot.drone.groundsdk.device.instrument.BatteryInfo
import com.parrot.drone.groundsdk.device.instrument.Gps
import com.parrot.drone.groundsdk.device.pilotingitf.Activable
import com.parrot.drone.groundsdk.device.pilotingitf.ManualCopterPilotingItf
import com.parrot.drone.groundsdk.facility.AutoConnection
import dalvik.system.DexClassLoader
import edu.cmu.cs.dronebrain.impl.ParrotAnafi
import edu.cmu.cs.dronebrain.impl.DebugCloudlet
import edu.cmu.cs.dronebrain.interfaces.FlightScript
import edu.cmu.cs.dronebrain.interfaces.Platform
import java.io.*
import java.net.HttpURLConnection
import java.net.URL

class MainActivity : Activity() {

    /** GroundSdk instance. */
    private lateinit var groundSdk: GroundSdk

    // Drone:
    /** Current drone instance. */
    private var drone: Drone? = null
    /** Reference to the current drone state. */
    private var droneStateRef: Ref<DeviceState>? = null
    /** Reference to the current drone battery info instrument. */
    private var droneBatteryInfoRef: Ref<BatteryInfo>? = null
    /** Reference to a current drone piloting interface. */
    private var pilotingItfRef: Ref<ManualCopterPilotingItf>? = null

    // Remote control:
    /** Current remote control instance. */
    private var rc: RemoteControl? = null
    /** Reference to the current remote control state. */
    private var rcStateRef: Ref<DeviceState>? = null
    /** Reference to the current remote control battery info instrument. */
    private var rcBatteryInfoRef: Ref<BatteryInfo>? = null

    // User Interface:
    /** Drone state text view. */
    private lateinit var droneStateTxt: TextView
    /** Drone battery level text view. */
    private lateinit var droneBatteryTxt: TextView
    /** Remote state level text view. */
    private lateinit var rcStateTxt: TextView
    /** Remote battery level text view. */
    private lateinit var rcBatteryTxt: TextView
    /** Take off / land button. */
    private lateinit var takeOffLandBt: Button
    /** Network for using LTE **/
    private lateinit var network: Network
    /** URL connection to transponder over LTE **/
    private lateinit var con: HttpURLConnection

    val TAG = "TAG"

    @Throws(Exception::class)
    fun getDrone(platform: Platform) : edu.cmu.cs.dronebrain.interfaces.Drone? {
        return if (platform == Platform.ANAFI) {
            ParrotAnafi()
        } else {
            throw Exception("Unsupported drone platfrom")
        }
    }

    fun getCloudlet() : edu.cmu.cs.dronebrain.interfaces.Cloudlet {
        return DebugCloudlet()
    }

    fun download(url: String?, dex: File?) {
        var bis: BufferedInputStream? = null
        var bos: BufferedOutputStream? = null
        try {
            val urlObject = URL(url)
            val httpConn = urlObject.openConnection() as HttpURLConnection
            val responseCode = httpConn.responseCode
            if (responseCode != HttpURLConnection.HTTP_OK) {
                throw RuntimeException("Download didn't work, http status: $responseCode")
            }
            bis = BufferedInputStream(httpConn.inputStream)
            bos = BufferedOutputStream(FileOutputStream(dex))
            copy(bis, bos)
        } catch (e: IOException) {
            throw RuntimeException(e)
        } finally {
            //closeQuietly(bos!!, bis!!)
        }
    }

    @Throws(IOException::class)
    private fun copy(src: BufferedInputStream, dst: OutputStream) {
        var len: Int
        val buf = ByteArray(8 * 1024)
        while (src.read(buf, 0, 8 * 1024).also { len = it } > 0) {
            dst.write(buf, 0, len)
        }
    }

    private fun closeQuietly(vararg cls: Closeable) {
        for (cl in cls) {
            if (cl != null) {
                try {
                    cl.close()
                } catch (ignored: IOException) {
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val policy = ThreadPolicy.Builder().permitAll().build()
        StrictMode.setThreadPolicy(policy)

        // Get user interface instances.
        droneStateTxt = findViewById(R.id.droneStateTxt)
        droneBatteryTxt = findViewById(R.id.droneBatteryTxt)
        rcStateTxt = findViewById(R.id.rcStateTxt)
        rcBatteryTxt = findViewById(R.id.rcBatteryTxt)
        takeOffLandBt = findViewById(R.id.takeOffLandBt)
        takeOffLandBt.setOnClickListener {onTakeOffLandClick()}

        // Initialize user interface default values.
        droneStateTxt.text = DeviceState.ConnectionState.DISCONNECTED.toString()
        rcStateTxt.text = DeviceState.ConnectionState.DISCONNECTED.toString()

        // Get a GroundSdk session.
        groundSdk = ManagedGroundSdk.obtainSession(this)
        // All references taken are linked to the activity lifecycle and
        // automatically closed at its destruction.

        val context = applicationContext
        val cm = context.getSystemService(
            CONNECTIVITY_SERVICE) as ConnectivityManager
        val req = NetworkRequest.Builder()
        req.addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
        cm.requestNetwork(req.build(), object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(net: Network) {
                network = net
                val url = URL("http://steel-eagle-dashboard.pgh.cloudapp.azurelel.cs.cmu.edu:8080/update")
                con = net.openConnection(url) as HttpURLConnection
                con.requestMethod = "POST"
                con.setRequestProperty("Content-Type", "application/json; utf-8")
                con.setRequestProperty("Accept", "application/json")
                con.doOutput = true
            } // Be sure to override other options in NetworkCallback() too...
        }
        )

        val f = File(getDir("dex", MODE_PRIVATE), "classes.dex")
        download("http://cloudlet040.elijah.cs.cmu.edu/classes.dex", f)

        val classLoader = DexClassLoader(
            f.absolutePath, getDir("outdex", MODE_PRIVATE).absolutePath, null,
            classLoader
        )

        try {
            val clazz = classLoader.loadClass("edu.cmu.cs.dronebrain.MS")
            Log.i(TAG, clazz.toString())
            Log.i(TAG, clazz.canonicalName)
            val inst = clazz.newInstance() as FlightScript
            Log.i(TAG, inst.toString())
            Log.i(TAG, inst.toString())
            val drone = getDrone(inst.platform) // Get the corresponding drone
            val cloudlet = getCloudlet() // Get the corresponding cloudlet
            inst.run(drone, cloudlet);
        } catch (e: java.lang.Exception) {
            Log.e(TAG, e.toString())
        }
    }

    override fun onStart() {
        super.onStart()

        // Monitor the auto connection facility.
        groundSdk.getFacility(AutoConnection::class.java) {
            // Called when the auto connection facility is available and when it changes.

            it?.let{
                // Start auto connection.
                if (it.status != AutoConnection.Status.STARTED) {
                    it.start()
                }

                // If the drone has changed.
                if (drone?.uid != it.drone?.uid) {
                    if(drone != null) {
                        // Stop monitoring the old drone.
                        stopDroneMonitors()

                        // Reset user interface drone part.
                        resetDroneUi()
                    }

                    // Monitor the new drone.
                    drone = it.drone
                    if(drone != null) {
                        startDroneMonitors()
                    }
                }

                // If the remote control has changed.
                if (rc?.uid  != it.remoteControl?.uid) {
                    if(rc != null) {
                        // Stop monitoring the old remote.
                        stopRcMonitors()

                        // Reset user interface Remote part.
                        resetRcUi()
                    }

                    // Monitor the new remote.
                    rc = it.remoteControl
                    if(rc != null) {
                        startRcMonitors()
                    }
                }
            }
        }
    }

    /**
     * Resets drone user interface part.
     */
    private fun resetDroneUi() {
        // Reset drone user interface views.
        droneStateTxt.text = DeviceState.ConnectionState.DISCONNECTED.toString()
        droneBatteryTxt.text = ""
        takeOffLandBt.isEnabled = false
    }

    /**
     * Starts drone monitors.
     */
    private fun startDroneMonitors() {
        // Monitor drone state.
        monitorDroneState()

        // Monitor drone battery level.
        monitorDroneBatteryLevel()

        // Monitor piloting interface.
        monitorPilotingInterface()

        transponder()
    }

    /**
     * Stops drone monitors.
     */
    private fun stopDroneMonitors() {
        // Close all references linked to the current drone to stop their monitoring.

        droneStateRef?.close()
        droneStateRef = null

        droneBatteryInfoRef?.close()
        droneBatteryInfoRef = null

        pilotingItfRef?.close()
        pilotingItfRef = null
    }

    /**
     * Monitor current drone state.
     */
    private fun monitorDroneState() {
        // Monitor current drone state.
        droneStateRef = drone?.getState {
            // Called at each drone state update.

            it?.let {
                // Update drone connection state view.
                droneStateTxt.text = it.connectionState.toString()
            }
        }
    }

    /**
     * Monitors current drone battery level.
     */
    private fun monitorDroneBatteryLevel() {
        // Monitor the battery info instrument.
        droneBatteryInfoRef = drone?.getInstrument(BatteryInfo::class.java) {
            // Called when the battery info instrument is available and when it changes.

            it?.let {
                // Update drone battery level view.
                droneBatteryTxt.text = "${it.batteryLevel}"
            }
        }
    }

    /**
     * Monitors current drone piloting interface.
     */
    private fun monitorPilotingInterface() {
        // Monitor a piloting interface.
        pilotingItfRef = drone?.getPilotingItf(ManualCopterPilotingItf::class.java) {
            // Called when the manual copter piloting Interface is available
            // and when it changes.

            // Disable the button if the piloting interface is not available.
            if (it == null) {
                takeOffLandBt.isEnabled = false
            } else {
                managePilotingItfState(it)
            }
        }
    }

    /**
     * Transponder
     */
    private fun transponder() {
        var lat : Double?
        var lon : Double?
        var alt : Double?
        // Monitor the GPS interface
        drone?.getInstrument(Gps::class.java) {
            it?.let {
                lat = it.lastKnownLocation()?.latitude
                lon = it.lastKnownLocation()?.longitude
                alt = it.lastKnownLocation()?.altitude
                Log.d("Drone", "Updating location $lat, $lon, $alt")
                if (lat != null && lon != null && alt != null) {
                    /** Make POST request with new data */
                    try {
                        var jsonInputString =
                            "{\"data\": {\"tag\": \"Anafi\", \"lat\":$lat, \"lng\":$lon, \"alt\":$alt, \"spd\":0, \"state\":\"Flying\", \"plan\":[]}, \"droneid\":\"Anafi\"}"
                        try {
                            val outputStream = DataOutputStream(con.outputStream)
                            outputStream.write(jsonInputString.toByteArray(charset("utf-8")))
                            outputStream.flush()
                        } catch (exception: Exception) {
                        }
                        Log.v("Drone", con.responseCode.toString())
                        val inputStream = DataInputStream(con.inputStream)
                        val reader: BufferedReader = BufferedReader(InputStreamReader(inputStream))
                        val output: String = reader.readLine()
                        Log.d("Drone", jsonInputString)
                    }
                    catch (e : Exception) {
                        Log.d("Exception", e.toString())
                    }
                }
            }
        }
    }


    /**
     * Manage piloting interface state.
     *
     * @param itf the piloting interface
     */
    private fun managePilotingItfState(itf: ManualCopterPilotingItf) {
        when(itf.state) {
            Activable.State.UNAVAILABLE -> {
                // Piloting interface is unavailable.
                takeOffLandBt.isEnabled = false
            }

            Activable.State.IDLE -> {
                // Piloting interface is idle.
                takeOffLandBt.isEnabled = false

                // Activate the interface.
                itf.activate()
            }

            Activable.State.ACTIVE -> {
                // Piloting interface is active.

                when {
                    itf.canTakeOff() -> {
                        // Drone can take off.
                        takeOffLandBt.isEnabled = true
                        takeOffLandBt.text = "Take Off"
                    }
                    itf.canLand() -> {
                        // Drone can land.
                        takeOffLandBt.isEnabled = true
                        takeOffLandBt.text = "Land"
                    }
                    else -> // Disable the button.
                        takeOffLandBt.isEnabled = false
                }
            }
        }
    }

    /**
     * Called on take off/land button click.
     */
    private fun onTakeOffLandClick() {
        // Get the piloting interface from its reference.
        pilotingItfRef?.get()?.let { itf ->
            // Do the action according to the interface capabilities
            if (itf.canTakeOff()) {
                // Take off
                itf.takeOff()
            } else if (itf.canLand()) {
                // Land
                itf.land()
            }
        }
    }

    /**
     * Resets remote user interface part.
     */
    private fun resetRcUi() {
        // Reset remote control user interface views.
        rcStateTxt.text = DeviceState.ConnectionState.DISCONNECTED.toString()
        rcBatteryTxt.text = ""
    }

    /**
     * Starts remote control monitors.
     */
    private fun startRcMonitors() {
        // Monitor remote state
        monitorRcState()

        // Monitor remote battery level
        monitorRcBatteryLevel()
    }

    /**
     * Stops remote control monitors.
     */
    private fun stopRcMonitors() {
        // Close all references linked to the current remote to stop their monitoring.

        rcStateRef?.close()
        rcStateRef = null

        rcBatteryInfoRef?.close()
        rcBatteryInfoRef = null
    }

    /**
     * Monitor current remote control state.
     */
    private fun monitorRcState() {
        // Monitor current drone state.
        rcStateRef = rc?.getState {
            // Called at each remote state update.

            it?.let {
                // Update remote connection state view.
                rcStateTxt.text = it.connectionState.toString()
            }
        }
    }

    /**
     * Monitors current remote control battery level.
     */
    private fun monitorRcBatteryLevel() {
        // Monitor the battery info instrument.
        rcBatteryInfoRef = rc?.getInstrument(BatteryInfo::class.java) {
            // Called when the battery info instrument is available and when it changes.

            it?.let {
                // Update drone battery level view.
                rcBatteryTxt.text = "${it.batteryLevel}"
            }
        }
    }
}
